---
title: "MSR-seq to bedRMod format"
author: "Luke R Frietze"
date: "2/12/2026"
output: html_document
editor_options: 
  chunk_output_type: console
---

#READ ME
```{r}
#Script: MSRseq_to_bedRmodv2_conversion.Rmd
# Description:
# Converts modification data generated by RNome_MSRseq_tRNAmod.Rmd, in this repository as data_cleaned_5_HRPC.csv.zip, into bedRmod v2 format.v Calculates modification confidence scores using a stoichiometry-based binomial model.

# Score calculation:
# Uses bedrmod_score_stoichiometry.R script originally developed by Iourii Motorine, Universit√© de Lorraine and adapted for MSRseq modification scoring by Luke R Frietze, University of Chicago.

# Built with:
#   R version 4.2.2 (2022-10-31)

# Package versions:
#   tidyverse 2.0.0
#   dplyr      1.1.3
#   openxlsx   4.2.5.2
#   Biostrings 2.64.1

# Input:
#   Modification data output from RNome_MSRseq_tRNAmod.Rmd in this repository as data_cleaned_5_HRPC.csv.zip, hg38_chromosomal_tRNA_genes_high_confidence_intro_remove+CCA_upper_case2.fa a reference sequence with the crhromosomal information for all positions and Table S3_sequence and modification sites for Fig3a.xlsx an exel with all known tRNA modifications MSRseq can detect for the most abundant tRNA isodecoders in HEK293T cells taken from Assari, M., B.J Chew, M.A. Bayat Tork, M. Sobczyk, C-Y. Zhu, H.A.V. Tran, D. Rudzka, W. Zhang, T. Pan: Decoding human tRNA modifications and crosstalk by enhanced single-read analysis, Genome Biol. in press (2026).

# Output:
#   bedRmod v2 formatted file with modification scores for each replicate and treatment in the original data. This is saved into the folder bedrmod

# Author:
#   Luke R Frietze

# Date:
#   2026-2-12

# ============================================================

```


#Set up packages
```{r, set up packages etc., echo=FALSE, results="hide", warning=FALSE, message=FALSE}
#Import packages 
library(tidyverse)
library(dplyr)
library(openxlsx)
library(Biostrings)

#import score calculating script
source("scripts p-value calculation/bedrmod_score_stoichiometry.R")

#Define a function that excludes things
`%nin%` = Negate(`%in%`)

empty_graph <- ggplot() + theme_void()
```


#rep 1
```{r}
#read modification excel data
modification_data <- read.csv(
  unz("data_cleaned_5_HRPC.csv.zip", "data_cleaned_5_HRPC.csv")
) 

#filter for needed data and calulate deletion rate and raw mutation rates
modification_data_fil <- modification_data %>%
  filter(pileup >= 50) %>%
  filter(bin_start == "60") %>%
  mutate(deletion_rate = deletion/pileup) %>%
  filter(rep == 1) %>%
  mutate(mutation_raw = round(pileup * mutation)) %>%
  select(-X,-T,-C,-G,-A,-N,-T, -stop,-insertion,-bin_stop,-bin_start, -rep)
  
#calculate score for mutation rates
mod_mut_score <- modification_data_fil %>%
  dplyr::rename(n_mod = mutation_raw,
         n_total = pileup)
mod_data_scores_score <- calculate_bedrmod_score_stoichiometry(mod_mut_score) %>%
  dplyr::rename(mut_score = score)

mut_scores_min <- mod_data_scores_score %>%
  select(gene, position, treatment, mut_score)

#calculate score for deletion rates
mod_del_score <- modification_data_fil %>%
  dplyr::rename(n_mod = deletion,
         n_total = pileup)
del_data_scores_score <- calculate_bedrmod_score_stoichiometry(mod_del_score) %>%
  dplyr::rename(del_score = score)

del_scores_min <- del_data_scores_score %>%
  select(gene, position, treatment, del_score)

#merge data frames together so has all modification information & filter for only ones greater than 5% modification
final_data <- modification_data_fil %>%
  left_join(mut_scores_min,
            by = c("gene", "position", "treatment")) %>%
  left_join(del_scores_min,
            by = c("gene", "position", "treatment")) %>%
  filter(mutation >= .05 | deletion_rate >= .05)
  


#read in reference sequence for chromosome data
ref_seq <- read.table(
  "hg38_chromosomal_tRNA_genes_high_confidence_intro_remove+CCA_upper_case2.fa",
  sep = "",
  stringsAsFactors = FALSE,
  fill = TRUE
) %>%
  as_tibble() %>%  
  filter(
    startsWith(V1, ">Homo_sapiens_tRNA")
  ) %>%
  mutate(V1 = sub(".*?-", "", V1)) %>%
  select(-V2,-V3,-V4,-V5,-V6,-V8,-V9,-V13) %>%
  separate(
    V11,
    into = c("chrom", "pos"),
    sep = ":"
  ) %>%
  separate(
    pos,
    into = c("chromStart_gene", "chromEnd_gene"),
    sep = "-"
  )  %>%
  dplyr::rename(
    gene       = V1,
    bp         = V7,
    scan_score = V10,
    strand     = V12
  )

#merge by gene
df_merged <- left_join(final_data, ref_seq, by = "gene")

# define for below to convert one letter codes to mod name
mod_map <- c(
  `"` = "m1A",
  I   = "I",
  O   = "m1I",
  M   = "ac4C",
  `'` = "m3C",
  `>` = "f5C",
  K   = "m1G",
  L   = "m2G",
  `#` = "Gm",
  R   = "m2,2G",
  `7` = "m7G",
  Q   = "Q",
  W   = "o2yW",
  `2` = "s2U",
  X   = "acp3U",
  P   = "Y"
)

#get modification positions and clean table to make usable
mod_positions <- read.xlsx("Table S3_sequence and modification sites for Fig3a.xlsx") %>%
  dplyr::rename(gene = Anticodon) %>%
  dplyr::select(gene, sequence) %>%
    mutate(
    gene = paste0(
      substr(gene, 1, 3), "-",        
      substr(gene, 4, nchar(gene)),   
      "-1"                            
    )
  ) %>%
   mutate(
    pos = str_split(sequence, "", simplify = TRUE) %>% as.data.frame()
  ) %>%
  unnest_wider(pos) %>%
  dplyr::rename_with(~ gsub("^pos\\$V", "", .x), -c(gene, sequence)) %>%
  dplyr::rename_with(~ gsub("V", "", .x), -c(gene, sequence)) %>%
  dplyr::select(-sequence) %>%
    pivot_longer(
    cols = -gene,
    names_to = "position",
    values_to = "nucleotide"
  ) %>%
mutate(
    position   = as.integer(position),
    nucleotide = na_if(nucleotide, "")
  ) %>%
  na.omit() %>%
  mutate(nucleotide = dplyr::recode(nucleotide, !!!mod_map)) %>%
  mutate(gene = ifelse(gene == "Sec-TCA-1-1", "SeC-TCA-1-1", gene),
         gene = ifelse(gene == "iMe-tCAT-1-1", "iMet-CAT-1-1", gene))

#join and clean more
df_all <- left_join(mod_positions, df_merged, by = c("gene", "position")) %>%
  na.omit() %>%
  filter(!nucleotide %in% c("A", "G", "U", "C")) %>%
  dplyr::rename(name = nucleotide) %>%
  mutate(chrom = gsub("chr", "", chrom)) %>%
  mutate(itemRgb = '0,0,0') %>%
  dplyr::rename(coverage = pileup) %>%
  mutate(chromStart = as.numeric(chromStart_gene) + as.numeric(position) - 1,
         chromEnd = as.numeric(chromStart_gene) + as.numeric(position)) %>%
  dplyr::select(-chromEnd_gene, -chromStart_gene, -source, -AA, -anticodon, -file_name) %>%
  mutate(thickStart = chromStart,
         thickEnd = chromEnd,
         score = coverage) %>%
  select(chrom, chromStart, chromEnd, name, mut_score, del_score, strand, thickStart, thickEnd, itemRgb, coverage, mutation, deletion_rate, treatment)

#make data frame for only BS treatment
bseq <- df_all %>%
  filter(treatment == "HRPC_BS",
          deletion_rate >= .05) %>%
  dplyr::select(-mut_score, -mutation, -treatment) %>%
  dplyr::rename(score = del_score,
                frequency = deletion_rate) %>%
  mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )

#make data frame for only CBH treatment
cbhseq <- df_all %>%
  filter(treatment == "HRPC_CBH",
          mutation >= .05) %>%
  dplyr::select(-del_score, -deletion_rate, -treatment) %>%
  dplyr::rename(score = mut_score,
                frequency = mutation) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )


#make data frame for only ctrl treatment mutation
ctrlseq_mut <- df_all %>%
  filter(treatment == "HRPC_ctrl",
          mutation >= .05) %>%
  dplyr::select(-del_score, -deletion_rate, -treatment) %>%
  dplyr::rename(score = mut_score,
                frequency = mutation) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )


#make data frame for only ctrl treatment deletion
ctrlseq_del <- df_all %>%
  filter(treatment == "HRPC_ctrl",
          deletion_rate >= .05) %>%
  dplyr::select(-mut_score, -mutation, -treatment) %>%
  dplyr::rename(score = del_score,
                frequency = deletion_rate) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )

#define function to make bedRModv2 files
write_bedrmod <- function(df, file) {
  
  metadata <- c(
    "#fileformat=bedRModv2",
    "#organism=9606",
    "#modification_type=RNA",
    "#modification_names=m1A:m1A:A,I:I:A,m1I:m1I:A,ac4C:ac4C:C,m3C:m3C:C,f5C:f5C:C,m1G:m1G:G,m2G:m2G:G,Gm:Gm:G,m22G:m22G:G,m7G:m7G:G,Q:Q:G,o2yW:o2yW:G,s2U:s2U:U,acp3U:acp3U:U,Y:Y:U",
    "#assembly=GRCh38",
    "#annotation_source=tRNAscan-SE",
    "#annotation_version=1",
    "#sequencing_platform=Illumina NovaSeqX",
    "#basecalling=",
    "#bioinformatics_workflow=https://github.com/Luke-F1875/MSRseq_data_processing_pipeline",
    "#experiment=https://doi.org/10.1038/s41467-022-30261-3",
    "#external_source="
  )

  # create commented header
  col_header <- paste0("#", paste(colnames(df), collapse = "\t"))

  # write header
  writeLines(c(metadata, col_header), con = file)

  # append data
  write.table(
    df,
    file = file,
    sep = "\t",
    quote = FALSE,
    row.names = FALSE,
    col.names = FALSE,
    append = TRUE
  )
}
#write the bedRModv2 files
write_bedrmod(bseq, "bedrmod/HRPC_BS_1.bedRMod")
write_bedrmod(cbhseq, "bedrmod/HRPC_CBH_1.bedRMod")
write_bedrmod(ctrlseq_mut, "bedrmod/HRPC_ctrl_mut_1.bedRMod")
write_bedrmod(ctrlseq_del, "bedrmod/HRPC_ctrl_del_1.bedRMod")

```


#rep 2
```{r}
#read modification excel data
modification_data <- read.csv(
  unz("data_cleaned_5_HRPC.csv.zip", "data_cleaned_5_HRPC.csv")
) 

#filter for needed data and calulate deletion rate and raw mutation rates
modification_data_fil <- modification_data %>%
  filter(pileup >= 50) %>%
  filter(bin_start == "60") %>%
  mutate(deletion_rate = deletion/pileup) %>%
  filter(rep == 2) %>%
  mutate(mutation_raw = round(pileup * mutation)) %>%
  select(-X,-T,-C,-G,-A,-N,-T, -stop,-insertion,-bin_stop,-bin_start, -rep)
  
#calculate score for mutation rates
mod_mut_score <- modification_data_fil %>%
  dplyr::rename(n_mod = mutation_raw,
         n_total = pileup)
mod_data_scores_score <- calculate_bedrmod_score_stoichiometry(mod_mut_score) %>%
  dplyr::rename(mut_score = score)

mut_scores_min <- mod_data_scores_score %>%
  select(gene, position, treatment, mut_score)

#calculate score for deletion rates
mod_del_score <- modification_data_fil %>%
  dplyr::rename(n_mod = deletion,
         n_total = pileup)
del_data_scores_score <- calculate_bedrmod_score_stoichiometry(mod_del_score) %>%
  dplyr::rename(del_score = score)

del_scores_min <- del_data_scores_score %>%
  select(gene, position, treatment, del_score)

#merge data frames together so has all modification information & filter for only ones greater than 5% modification
final_data <- modification_data_fil %>%
  left_join(mut_scores_min,
            by = c("gene", "position", "treatment")) %>%
  left_join(del_scores_min,
            by = c("gene", "position", "treatment")) %>%
  filter(mutation >= .05 | deletion_rate >= .05)
  


#read in reference sequence for chromosome data
ref_seq <- read.table(
  "hg38_chromosomal_tRNA_genes_high_confidence_intro_remove+CCA_upper_case2.fa",
  sep = "",
  stringsAsFactors = FALSE,
  fill = TRUE
) %>%
  as_tibble() %>%  
  filter(
    startsWith(V1, ">Homo_sapiens_tRNA")
  ) %>%
  mutate(V1 = sub(".*?-", "", V1)) %>%
  select(-V2,-V3,-V4,-V5,-V6,-V8,-V9,-V13) %>%
  separate(
    V11,
    into = c("chrom", "pos"),
    sep = ":"
  ) %>%
  separate(
    pos,
    into = c("chromStart_gene", "chromEnd_gene"),
    sep = "-"
  )  %>%
  dplyr::rename(
    gene       = V1,
    bp         = V7,
    scan_score = V10,
    strand     = V12
  )

#merge by gene
df_merged <- left_join(final_data, ref_seq, by = "gene")

# define for below to convert one letter codes to mod name
mod_map <- c(
  `"` = "m1A",
  I   = "I",
  O   = "m1I",
  M   = "ac4C",
  `'` = "m3C",
  `>` = "f5C",
  K   = "m1G",
  L   = "m2G",
  `#` = "Gm",
  R   = "m2,2G",
  `7` = "m7G",
  Q   = "Q",
  W   = "o2yW",
  `2` = "s2U",
  X   = "acp3U",
  P   = "Y"
)

#get modification positions and clean table to make usable
mod_positions <- read.xlsx("Table S3_sequence and modification sites for Fig3a.xlsx") %>%
  dplyr::rename(gene = Anticodon) %>%
  dplyr::select(gene, sequence) %>%
    mutate(
    gene = paste0(
      substr(gene, 1, 3), "-",        
      substr(gene, 4, nchar(gene)),   
      "-1"                            
    )
  ) %>%
   mutate(
    pos = str_split(sequence, "", simplify = TRUE) %>% as.data.frame()
  ) %>%
  unnest_wider(pos) %>%
  dplyr::rename_with(~ gsub("^pos\\$V", "", .x), -c(gene, sequence)) %>%
  dplyr::rename_with(~ gsub("V", "", .x), -c(gene, sequence)) %>%
  dplyr::select(-sequence) %>%
    pivot_longer(
    cols = -gene,
    names_to = "position",
    values_to = "nucleotide"
  ) %>%
mutate(
    position   = as.integer(position),
    nucleotide = na_if(nucleotide, "")
  ) %>%
  na.omit() %>%
  mutate(nucleotide = dplyr::recode(nucleotide, !!!mod_map)) %>%
  mutate(gene = ifelse(gene == "Sec-TCA-1-1", "SeC-TCA-1-1", gene),
         gene = ifelse(gene == "iMe-tCAT-1-1", "iMet-CAT-1-1", gene))

#join and clean more
df_all <- left_join(mod_positions, df_merged, by = c("gene", "position")) %>%
  na.omit() %>%
  filter(!nucleotide %in% c("A", "G", "U", "C")) %>%
  dplyr::rename(name = nucleotide) %>%
  mutate(chrom = gsub("chr", "", chrom)) %>%
  mutate(itemRgb = '0,0,0') %>%
  dplyr::rename(coverage = pileup) %>%
  mutate(chromStart = as.numeric(chromStart_gene) + as.numeric(position) - 1,
         chromEnd = as.numeric(chromStart_gene) + as.numeric(position)) %>%
  dplyr::select(-chromEnd_gene, -chromStart_gene, -source, -AA, -anticodon, -file_name) %>%
  mutate(thickStart = chromStart,
         thickEnd = chromEnd,
         score = coverage) %>%
  select(chrom, chromStart, chromEnd, name, mut_score, del_score, strand, thickStart, thickEnd, itemRgb, coverage, mutation, deletion_rate, treatment)

#make data frame for only BS treatment
bseq <- df_all %>%
  filter(treatment == "HRPC_BS",
          deletion_rate >= .05) %>%
  dplyr::select(-mut_score, -mutation, -treatment) %>%
  dplyr::rename(score = del_score,
                frequency = deletion_rate) %>%
  mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )

#make data frame for only CBH treatment
cbhseq <- df_all %>%
  filter(treatment == "HRPC_CBH",
          mutation >= .05) %>%
  dplyr::select(-del_score, -deletion_rate, -treatment) %>%
  dplyr::rename(score = mut_score,
                frequency = mutation) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )


#make data frame for only ctrl treatment mutation
ctrlseq_mut <- df_all %>%
  filter(treatment == "HRPC_ctrl",
          mutation >= .05) %>%
  dplyr::select(-del_score, -deletion_rate, -treatment) %>%
  dplyr::rename(score = mut_score,
                frequency = mutation) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )


#make data frame for only ctrl treatment deletion
ctrlseq_del <- df_all %>%
  filter(treatment == "HRPC_ctrl",
          deletion_rate >= .05) %>%
  dplyr::select(-mut_score, -mutation, -treatment) %>%
  dplyr::rename(score = del_score,
                frequency = deletion_rate) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )

#define function to make bedRModv2 files
write_bedrmod <- function(df, file) {
  
  metadata <- c(
    "#fileformat=bedRModv2",
    "#organism=9606",
    "#modification_type=RNA",
    "#modification_names=m1A:m1A:A,I:I:A,m1I:m1I:A,ac4C:ac4C:C,m3C:m3C:C,f5C:f5C:C,m1G:m1G:G,m2G:m2G:G,Gm:Gm:G,m22G:m22G:G,m7G:m7G:G,Q:Q:G,o2yW:o2yW:G,s2U:s2U:U,acp3U:acp3U:U,Y:Y:U",
    "#assembly=GRCh38",
    "#annotation_source=tRNAscan-SE",
    "#annotation_version=1",
    "#sequencing_platform=Illumina NovaSeqX",
    "#basecalling=",
    "#bioinformatics_workflow=https://github.com/Luke-F1875/MSRseq_data_processing_pipeline",
    "#experiment=https://doi.org/10.1038/s41467-022-30261-3",
    "#external_source="
  )

  # create commented header
  col_header <- paste0("#", paste(colnames(df), collapse = "\t"))

  # write header
  writeLines(c(metadata, col_header), con = file)

  # append data
  write.table(
    df,
    file = file,
    sep = "\t",
    quote = FALSE,
    row.names = FALSE,
    col.names = FALSE,
    append = TRUE
  )
}
#write the bedRModv2 files
write_bedrmod(bseq, "bedrmod/HRPC_BS_2.bedRMod")
write_bedrmod(cbhseq, "bedrmod/HRPC_CBH_2.bedRMod")
write_bedrmod(ctrlseq_mut, "bedrmod/HRPC_ctrl_mut_2.bedRMod")
write_bedrmod(ctrlseq_del, "bedrmod/HRPC_ctrl_del_2.bedRMod")

```


#rep 3
```{r}
#read modification excel data
modification_data <- read.csv(
  unz("data_cleaned_5_HRPC.csv.zip", "data_cleaned_5_HRPC.csv")
) 

#filter for needed data and calulate deletion rate and raw mutation rates
modification_data_fil <- modification_data %>%
  filter(pileup >= 50) %>%
  filter(bin_start == "60") %>%
  mutate(deletion_rate = deletion/pileup) %>%
  filter(rep == 3) %>%
  mutate(mutation_raw = round(pileup * mutation)) %>%
  select(-X,-T,-C,-G,-A,-N,-T, -stop,-insertion,-bin_stop,-bin_start, -rep)
  
#calculate score for mutation rates
mod_mut_score <- modification_data_fil %>%
  dplyr::rename(n_mod = mutation_raw,
         n_total = pileup)
mod_data_scores_score <- calculate_bedrmod_score_stoichiometry(mod_mut_score) %>%
  dplyr::rename(mut_score = score)

mut_scores_min <- mod_data_scores_score %>%
  select(gene, position, treatment, mut_score)

#calculate score for deletion rates
mod_del_score <- modification_data_fil %>%
  dplyr::rename(n_mod = deletion,
         n_total = pileup)
del_data_scores_score <- calculate_bedrmod_score_stoichiometry(mod_del_score) %>%
  dplyr::rename(del_score = score)

del_scores_min <- del_data_scores_score %>%
  select(gene, position, treatment, del_score)

#merge data frames together so has all modification information & filter for only ones greater than 5% modification
final_data <- modification_data_fil %>%
  left_join(mut_scores_min,
            by = c("gene", "position", "treatment")) %>%
  left_join(del_scores_min,
            by = c("gene", "position", "treatment")) %>%
  filter(mutation >= .05 | deletion_rate >= .05)
  


#read in reference sequence for chromosome data
ref_seq <- read.table(
  "hg38_chromosomal_tRNA_genes_high_confidence_intro_remove+CCA_upper_case2.fa",
  sep = "",
  stringsAsFactors = FALSE,
  fill = TRUE
) %>%
  as_tibble() %>%  
  filter(
    startsWith(V1, ">Homo_sapiens_tRNA")
  ) %>%
  mutate(V1 = sub(".*?-", "", V1)) %>%
  select(-V2,-V3,-V4,-V5,-V6,-V8,-V9,-V13) %>%
  separate(
    V11,
    into = c("chrom", "pos"),
    sep = ":"
  ) %>%
  separate(
    pos,
    into = c("chromStart_gene", "chromEnd_gene"),
    sep = "-"
  )  %>%
  dplyr::rename(
    gene       = V1,
    bp         = V7,
    scan_score = V10,
    strand     = V12
  )

#merge by gene
df_merged <- left_join(final_data, ref_seq, by = "gene")

# define for below to convert one letter codes to mod name
mod_map <- c(
  `"` = "m1A",
  I   = "I",
  O   = "m1I",
  M   = "ac4C",
  `'` = "m3C",
  `>` = "f5C",
  K   = "m1G",
  L   = "m2G",
  `#` = "Gm",
  R   = "m2,2G",
  `7` = "m7G",
  Q   = "Q",
  W   = "o2yW",
  `2` = "s2U",
  X   = "acp3U",
  P   = "Y"
)

#get modification positions and clean table to make usable
mod_positions <- read.xlsx("Table S3_sequence and modification sites for Fig3a.xlsx") %>%
  dplyr::rename(gene = Anticodon) %>%
  dplyr::select(gene, sequence) %>%
    mutate(
    gene = paste0(
      substr(gene, 1, 3), "-",        
      substr(gene, 4, nchar(gene)),   
      "-1"                            
    )
  ) %>%
   mutate(
    pos = str_split(sequence, "", simplify = TRUE) %>% as.data.frame()
  ) %>%
  unnest_wider(pos) %>%
  dplyr::rename_with(~ gsub("^pos\\$V", "", .x), -c(gene, sequence)) %>%
  dplyr::rename_with(~ gsub("V", "", .x), -c(gene, sequence)) %>%
  dplyr::select(-sequence) %>%
    pivot_longer(
    cols = -gene,
    names_to = "position",
    values_to = "nucleotide"
  ) %>%
mutate(
    position   = as.integer(position),
    nucleotide = na_if(nucleotide, "")
  ) %>%
  na.omit() %>%
  mutate(nucleotide = dplyr::recode(nucleotide, !!!mod_map)) %>%
  mutate(gene = ifelse(gene == "Sec-TCA-1-1", "SeC-TCA-1-1", gene),
         gene = ifelse(gene == "iMe-tCAT-1-1", "iMet-CAT-1-1", gene))

#join and clean more
df_all <- left_join(mod_positions, df_merged, by = c("gene", "position")) %>%
  na.omit() %>%
  filter(!nucleotide %in% c("A", "G", "U", "C")) %>%
  dplyr::rename(name = nucleotide) %>%
  mutate(chrom = gsub("chr", "", chrom)) %>%
  mutate(itemRgb = '0,0,0') %>%
  dplyr::rename(coverage = pileup) %>%
  mutate(chromStart = as.numeric(chromStart_gene) + as.numeric(position) - 1,
         chromEnd = as.numeric(chromStart_gene) + as.numeric(position)) %>%
  dplyr::select(-chromEnd_gene, -chromStart_gene, -source, -AA, -anticodon, -file_name) %>%
  mutate(thickStart = chromStart,
         thickEnd = chromEnd,
         score = coverage) %>%
  select(chrom, chromStart, chromEnd, name, mut_score, del_score, strand, thickStart, thickEnd, itemRgb, coverage, mutation, deletion_rate, treatment)

#make data frame for only BS treatment
bseq <- df_all %>%
  filter(treatment == "HRPC_BS",
          deletion_rate >= .05) %>%
  dplyr::select(-mut_score, -mutation, -treatment) %>%
  dplyr::rename(score = del_score,
                frequency = deletion_rate) %>%
  mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )

#make data frame for only CBH treatment
cbhseq <- df_all %>%
  filter(treatment == "HRPC_CBH",
          mutation >= .05) %>%
  dplyr::select(-del_score, -deletion_rate, -treatment) %>%
  dplyr::rename(score = mut_score,
                frequency = mutation) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )


#make data frame for only ctrl treatment mutation
ctrlseq_mut <- df_all %>%
  filter(treatment == "HRPC_ctrl",
          mutation >= .05) %>%
  dplyr::select(-del_score, -deletion_rate, -treatment) %>%
  dplyr::rename(score = mut_score,
                frequency = mutation) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )


#make data frame for only ctrl treatment deletion
ctrlseq_del <- df_all %>%
  filter(treatment == "HRPC_ctrl",
          deletion_rate >= .05) %>%
  dplyr::select(-mut_score, -mutation, -treatment) %>%
  dplyr::rename(score = del_score,
                frequency = deletion_rate) %>%
    mutate(frequency = frequency * 100) %>%
  dplyr::select(
    chrom,
    chromStart,
    chromEnd,
    name,
    score,
    strand,
    thickStart,
    thickEnd,
    itemRgb,
    coverage,
    frequency
  )

#define function to make bedRModv2 files
write_bedrmod <- function(df, file) {
  
  metadata <- c(
    "#fileformat=bedRModv2",
    "#organism=9606",
    "#modification_type=RNA",
    "#modification_names=m1A:m1A:A,I:I:A,m1I:m1I:A,ac4C:ac4C:C,m3C:m3C:C,f5C:f5C:C,m1G:m1G:G,m2G:m2G:G,Gm:Gm:G,m22G:m22G:G,m7G:m7G:G,Q:Q:G,o2yW:o2yW:G,s2U:s2U:U,acp3U:acp3U:U,Y:Y:U",
    "#assembly=GRCh38",
    "#annotation_source=tRNAscan-SE",
    "#annotation_version=1",
    "#sequencing_platform=Illumina NovaSeqX",
    "#basecalling=",
    "#bioinformatics_workflow=https://github.com/Luke-F1875/MSRseq_data_processing_pipeline",
    "#experiment=https://doi.org/10.1038/s41467-022-30261-3",
    "#external_source="
  )

  # create commented header
  col_header <- paste0("#", paste(colnames(df), collapse = "\t"))

  # write header
  writeLines(c(metadata, col_header), con = file)

  # append data
  write.table(
    df,
    file = file,
    sep = "\t",
    quote = FALSE,
    row.names = FALSE,
    col.names = FALSE,
    append = TRUE
  )
}
#write the bedRModv2 files
write_bedrmod(bseq, "bedrmod/HRPC_BS_3.bedRMod")
write_bedrmod(cbhseq, "bedrmod/HRPC_CBH_3.bedRMod")
write_bedrmod(ctrlseq_mut, "bedrmod/HRPC_ctrl_mut_3.bedRMod")
write_bedrmod(ctrlseq_del, "bedrmod/HRPC_ctrl_del_3.bedRMod")

```



